Add type filter for feed based on ap_object_type taxonomy

pfefferle opened the pull request:
## Summary
Adds a button-based filter to the ActivityPub feed that allows users to filter posts by object type (Article, Note, etc.) using the `ap_object_type` taxonomy.

## Changes
- Add `ap_object_type` field to `FeedPost` TypeScript interface
- Add `apObjectType` parameter to `useFeed` hook with REST API integration
- Create `TypeFilter` component using WordPress ButtonGroup
- Integrate TypeFilter into feed stage with localStorage persistence
- Filter buttons appear in DataViews header, aligned to the right

## Features
- **Button-based UI**: Clean ButtonGroup interface with "All", "Article", "Note" buttons
- **localStorage persistence**: Selected filter is remembered across browser sessions
- **REST API integration**: Properly filters posts via the `ap_object_type` taxonomy parameter
- **Reactive updates**: Feed updates immediately when filter selection changes

## Test plan
- [ ] View the feed in the ActivityPub Social Web reader
- [ ] Verify filter buttons appear in the header on the right side
- [ ] Click "Article" button and verify only Article posts are shown
- [ ] Click "Note" button and verify only Note posts are shown
- [ ] Click "All" button and verify all posts are shown
- [ ] Refresh the page and verify the selected filter persists
- [ ] Clear localStorage and verify it defaults to "All"

## Screenshots

<img width="1211" height="513" alt="Screenshot 2025-11-20 at 18 36 35" src="https://github.com/user-attachments/assets/eadb10bf-4996-4977-9699-bc585655c3eb" />

pfefferle committed:
Add type filter for feed based on ap_object_type taxonomy

This adds a button-based filter to the feed that allows users to filter posts by ActivityPub object type (Article, Note, etc.).

Changes:
- Add ap_object_type field to FeedPost interface
- Add apObjectType parameter to useFeed hook with REST API integration
- Create TypeFilter component using WordPress ButtonGroup
- Integrate TypeFilter into feed stage with localStorage persistence
- Filter buttons appear in DataViews header aligned to the right

The selected filter is persisted to localStorage so it's remembered across sessions.

pfefferle committed:
Merge branch 'add/reader' into add/feed-type-filter

obenland commented:
Nice! Is that something we could add to the DataView settings?

pfefferle commented:
I tried... will give it another 30 mins...

pfefferle committed:
Refactor: Use native DataViews filter for ap_object_type

Replace custom TypeFilter component with native DataViews filtering system.

Changes:
- Remove custom TypeFilter ButtonGroup component
- Add ap_object_type as a DataView filter field with 'isAny' operator
- Extract filter values from view.filters instead of separate state
- Remove localStorage persistence code (DataViews handles this internally)
- Taxonomy terms populate filter dropdown automatically

This integrates the type filter into DataViews' native filtering UI, providing:
- Consistent UX with other DataViews filters
- Built-in persistence through WordPress views system
- Better accessibility and keyboard navigation
- Automatic filter state management

pfefferle commented:
@obenland I am not sure if I like that more!? https://github.com/Automattic/wordpress-activitypub/pull/2510/commits/025afc94734e0ee833efbcde0a6b8a7d54664bc8

pfefferle commented:
Might be nicer if we have some more filters in the future (single actor feeds could maybe be handled as filter too)

obenland commented:
Hm, it looks like I don't have any articles…

obenland commented:
Are articles being associated correctly in the inbox? Did you create your test-articles manually?

I can't get the selector to show up, which has me wondering if that's because I only have note posts

pfefferle commented:
I have added some Articles, by hand and through the inbox.

obenland commented:
Hm, I do have one article post…

obenland commented:
oh! I was looking in the wrong place

pfefferle committed:
Merge branch 'add/reader' into add/feed-type-filter

pfefferle commented:
@obenland I just re-tested with my blogs and it works as expected.

pfefferle committed:
Add ActivityPub object type filter to feed

Introduces a filter for ActivityPub object types in the feed-stage.js component. Fetches available object types and allows filtering posts by type, updating the feed query and UI accordingly.

obenland commented on src/social-web/routes/feed/stage.tsx:
```diff
@@ -53,6 +54,11 @@ export default function FeedStage( { onSelectItem }: FeedStageProps ) {
 
+	// Fetch ap_object_type taxonomy terms for filter
+	const { records: apObjectTypes } = useEntityRecords( 'taxonomy', 'ap_object_type', {
+		per_page: 100,
```
We could probably drop that? I assume we'd only ever expect 2 types to be returned? Definitely less than 10?

We could also consider implementing the `elements` array as a static array to avoid this request and make the labels translatable.

obenland commented on src/social-web/routes/feed/stage.tsx:
```diff
@@ -125,18 +131,44 @@ export default function FeedStage( { onSelectItem }: FeedStageProps ) {
 		order: view.sort?.direction || 'desc',
 		search: view.search || '',
 		userId: activeActorId,
+		apObjectType: apObjectTypeFilter,
```
```suggestion
		filters: view.filters,
```

Should we pass all `view.filters` here and let `useFeed()` dissect them?

obenland commented on src/social-web/routes/feed/stage.tsx:
```diff
@@ -125,18 +131,44 @@ export default function FeedStage( { onSelectItem }: FeedStageProps ) {
+					label: term.name,
+				} ) ) || [],
+			filterBy: {
+				operators: [ 'isAny' ],
```
```suggestion
				operators: [ 'is' ],
```

We'd probably only want one to be selected at any time?

pfefferle replied:
I am not sure about that, but let's start with one

pfefferle replied:
It does have to be isAny... a taxonomy seem to not work otherwise:

> 'is' operator:
> - Used for single-value comparisons
>  - Matches when the field value exactly equals the filter value
>  - Best for boolean or single-select filters
>
>  'isAny' operator:
>  - Used for multi-value/enum comparisons
>  - Matches when the field value is any one of the selected filter values
>  - Required when you have elements defined (multiple options to choose from)
>  - Allows users to select multiple types (e.g., filter for "Article OR Note")
>
>  In your case, since you defined elements with multiple object types (Article, Note, etc.) and users can potentially select multiple types to filter by, WordPress DataViews expects the 'isAny' operator to
  properly:
>
>  1. Display the filter UI with checkboxes for multiple selections
>  2. Apply the filter logic correctly (show posts that match ANY of the selected types)
>  3. Store and persist the filter state properly
>
>  With 'is', DataViews likely didn't recognize it as a valid enum-style filter and either didn't display the filter UI at all or didn't persist the selection correctly. The 'isAny' operator tells DataViews: "this
>   field has predefined options (elements) and we want to filter by any matching value."

obenland commented on src/social-web/routes/feed/stage.tsx:
```diff
@@ -125,18 +131,44 @@ export default function FeedStage( { onSelectItem }: FeedStageProps ) {
+				} ) ) || [],
+			filterBy: {
+				operators: [ 'isAny' ],
+			},
```
```suggestion
				render: () => null,
			},
```
Claude felt like it was important to define this explicitly. Not sure if it actually is.

obenland commented on src/social-web/routes/feed/stage.tsx:
```diff
@@ -125,18 +131,44 @@ export default function FeedStage( { onSelectItem }: FeedStageProps ) {
+			},
+		} ),
+		[ apObjectTypes ]
+	);
```
Let's move `apObjectTypeField` into its own field component, alongside the others.

pfefferle committed:
Update src/social-web/routes/feed/stage.tsx

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update src/social-web/routes/feed/stage.tsx

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update src/social-web/routes/feed/stage.tsx

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Fix comment date display in Social Web feed inspector

Use date_gmt field instead of date for reliable UTC parsing in comment timestamps. This ensures comment dates display correctly using relative time format (5m, 2h, 6d) consistent with post dates.

pfefferle committed:
Merge branch 'add/reader' into add/feed-type-filter

pfefferle committed:
Refactor feed-stage.js for improved filter handling

Reworks the ActivityPub feed-stage.js to improve filter and query param handling, including more robust support for taxonomy filters and ap_object_type. Updates the filterBy operators and render logic for the 'Type' filter, and refactors state management for view updates and feed queries.

pfefferle committed:
Remove per_page param from ap_object_type taxonomy fetch

The ap_object_type taxonomy terms are now fetched without specifying the per_page parameter, relying on the default behavior of useEntityRecords. This simplifies the code and avoids unnecessary explicit pagination.

pfefferle committed:
Add object type filter to feed view

Introduces support for filtering ActivityPub feed posts by object type. Updates feed-stage.js, feed route, and related hooks to handle object type taxonomy, and adds a new object-type field component. Also updates post type registration to include the ap_object_type parameter.

obenland commented on includes/class-post-types.php:
```diff
@@ -706,6 +706,17 @@ function ( $params ) {
 				);
 
+				$params['ap_object_type'] = array(
+					'description'       => __( 'Filter posts by ActivityPub object type.', 'activitypub' ),
```
```suggestion
					'description'       => 'Filter posts by ActivityPub object type.',
```

obenland commented on includes/class-post-types.php:
```diff
@@ -706,6 +706,17 @@ function ( $params ) {
+					),
+					'sanitize_callback' => function ( $value ) {
+						return array_map( 'absint', (array) $value );
+					},
```
Is the built-in sanitization not taking care of this? We could set `'minimum' => 0` if we want to avoid negative numbers. Alternatively we could validate the passed value by checking if an ap_object_type term with that ID exists

obenland committed:
Refactor: Use native DataViews filter for ap_object_type

Refactors the object type filter to follow Gutenberg field patterns:

- Use async `getElements` to fetch taxonomy terms (no React hooks in field)
- Change to single-value filtering with `is` operator instead of array-based `isAny`
- Extract first value from ap_object_type array for cleaner UX
- Fix filter handling in useFeed to support both single values and arrays
- Remove unused useEntityRecords hook from feed stage
- Simplify field definition following @wordpress/fields architecture

This provides a cleaner, radio-style selection for filtering posts by type
(Article, Note, etc.) while maintaining compatibility with the taxonomy structure.

obenland reviewed (commented):
In bbfe23d0c002569e74d36c2c710de7b347a06c1a I moved all of the entity stuff into the field definition, too, and refactored it to enable single type selections, I hope that's okay. I meant to provide all that stuff as feedback but it got to be a lot and pretty intricate.

The main missing thing is making these term names more accessible. `Article` and `Note` are technical terms (we literally save what activities send in their type), and not necessarily the best way to describe what we're trying to convey. I'm also not sure they're translated or translatable?

pfefferle committed:
class-post-types.php aktualisieren

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland committed:
Update object type filter with user-friendly labels

Filter to only Note and Article types and display them as
"Short form" and "Long form" for better user comprehension.
Also adds proper TypeScript typing using Term from @wordpress/core-data.

obenland commented:
In b2d41d2c, updated to limit the filter to notes and articles and make the labels translatable. I'm not sure "Short form" and "Long form" are the best labels, happy for us to change those. It doesn't seem to be easily possible to not add the filter when there's only one term.

pfefferle commented:
I am not a fan of limiting the types! This is a similar pattern, Mastodon used (or still uses) an what made `Article`s a second class object on the Fediverse!

I will think about a way to better support all kinds of Objects and to translate at least most of them.

pfefferle committed:
Expand object type labels and options

Updated the object type field to support all ActivityPub object types with user-friendly labels, instead of only Note and Article. The feed now displays and filters by a broader set of types, improving usability and consistency.

pfefferle committed:
Add minimum value to ActivityPub object type filter

Set 'minimum' to 0 for integer items in the ActivityPub object type filter parameter and remove the custom sanitize_callback. This enforces non-negative values at the schema level.

pfefferle commented:
@obenland Added some small changes to have best of both worlds. I also changed the translatable text a bit... I think "Long form" and "Short form" are no common (end-user friendly) terms...
